### PR TITLE
ch4/ofi: implement native path rndv

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Based on #7183, this PR adds the ability for ch4:ofi netmod to switch to RNDV. Sender may send an empty message with specific protocol bits marked in the match_bits. Receiver will reply with its matched `rreq`, then both can continue with either `am_tag_{send,recv}` (this PR) or more complex protocols (future PRs).

#### Eager Send
image
#### Eager Ssend
image
#### RNDV Send

image

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi/more

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch4/ofi/more

**Review by raffenet — COMMENTED** [review]
I did a quick review to try and stay ahead of this one. There's one correctness issue a reference count. My other concern is about a TODO comment.

File: src/mpid/ch4/netmod/ofi/ofi_recv.h [hunk 1]
```diff
@@ -158,7 +158,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
     MPIDI_OFI_REQUEST(rreq, buf) = buf;
     MPIDI_OFI_REQUEST(rreq, count) = count;
     MPIDI_OFI_REQUEST(rreq, datatype) = datatype;
-    MPIR_Datatype_release_if_not_builtin(datatype);
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_recv.h:161]: Yeah, mixed up in fixup commits. I'll correct it.

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 2]
```diff
@@ -596,6 +607,32 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
 
         *request = MPIR_Request_create_complete(MPIR_REQUEST_KIND__SEND);
         MPIR_ERR_CHECK(mpi_errno);
+    } else if (syncflag && data_sz > MPIR_CVAR_CH4_OFI_EAGER_THRESHOLD) {
+        /* RNDV path - TODO: always do RNDV if (!is_am && data_sz > EAGER_THRESHOLD) */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:617]: The native EAGER will always be the default. However, will there be an advantage for native RNDV? I guess hardware progress will be one, but that will depend on the system, and users or facilities can always adjust the EAGER threshold for that. Anyway, the plan is to delay that discussion as a TODO for next release cycle.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:617]: A side note, I think we should provide users a single EAGER threshold that governs all paths. Most MPI users understand (can be taught about) EAGER vs RNDV; it is frustrating if there isn't a single button or the button does not do what users expect.

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/netmod/ofi/ofi_send.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:617]: > The native EAGER will always be the default. However, will there be an performance advantage for native RNDV? I guess hardware progress will be one, but that will depend on the system, and users or facilities can always adjust the EAGER threshold for that. Anyway, the plan is to delay that discussion as a TODO for next release cycle.

We did discuss this and I thought we were on the same page, but perhaps not. I will state again that if the goal is disabling provider-based rndv in favor of our software approach, I am against it.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:617]: Okay. Is this a future discussion or a current blocker?

[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:617]: The TODO comment suggests an intent that I am uncomfortable with. Please remove it and we can move forward with this PR.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h]: I see. No problem! Just updated the PR.

File: src/mpid/ch4/netmod/ofi/ofi_rndv.c [hunk 3]
```diff
@@ -131,9 +188,28 @@ int MPIDI_OFI_rndv_cts_event(int vci, struct fi_cq_tagged_entry *wc, MPIR_Reques
...
+        /* just reply with data_sz. Receiver will send another CTS */
+        struct rndv_info_hdr rndv_info;
+        rndv_info.sreq = sreq;
+        rndv_info.rreq = hdr->rreq;
+        MPIDI_Datatype_check_size(MPIDIG_REQUEST(sreq, datatype), MPIDIG_REQUEST(sreq, count),
+                                  rndv_info.data_sz);
+        mpi_errno = MPIDI_NM_am_send_hdr_reply(sreq->comm, MPIDIG_REQUEST(sreq, u.send.dest),
+                                               MPIDI_OFI_RNDV_INFO, &rndv_info, sizeof(rndv_info),
+                                               MPIDIG_REQUEST(sreq, req->local_vci),
+                                               MPIDIG_REQUEST(sreq, req->remote_vci));
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_rndv.c:207]: We don't have to handle it now, but this patch is making me reconsider whether we should use the user's buffer to store additional information for the RTS. It's just a matter of unpacking in the noncontig case, which seems preferable to more packets and handlers.

**Review by raffenet — CHANGES_REQUESTED** [review]

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 4]
```diff
@@ -37,6 +37,16 @@
...
 
+    - name        : MPIR_CVAR_CH4_OFI_EAGER_THRESHOLD
+      category    : CH4_OFI
+      type        : int
+      default     : 16384
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Message below MPIR_CVAR_CH4_OFI_EAGER_THRESHOLD will be sent eagerly.
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h]: Added a commit to change the default to -1 and updated the description text.

File: src/mpid/ch4/netmod/ofi/ofi_rndv.c (hunk 3 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_rndv.c:207]: Other than for testing purposes, the RNDV path should only be opted for large message or messages with significant latency where a couple protocol handshake does not meaningfully impact the overall performance. We are talking less than a couple microseconds for message latencies over 500 microseconds (or more depend on the threshold). On the other hand, double using using user buffers are quire complex. We need consider whether user buffer is big enough to receive the protocol header data; we need consider different receive mode (pack, iov). And if the recv buffer is direct GPU buffer, we need consider the cost of access latency. Thus, I think the simplicity of using additional handshake overwhelms the complexity of double using user buffers.

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 5]
```diff
@@ -37,6 +37,19 @@
...
+      category    : CH4_OFI
+      type        : int
+      default     : -1
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Messages below MPIR_CVAR_CH4_OFI_EAGER_THRESHOLD will be sent eagerly. Messages
+        above the threshold will perform redezvous handshake before sending the data.
+        If set to -1, the messages will be sent eagerly up to provider max_msg_size.
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_send.h:50]: I don't want to give the impression that messages up to max_msg_size are sent eagerly. How about:
```
Messages below MPIR_CVAR_CH4_OFI_EAGER_THRESHOLD will be sent eagerly using fi_tagged interfaces. Messages above the threshold will perform an MPICH-level rendezvous handshake before sending the data. If set to -1, MPICH will only perform rendezvous for messages larger than the provider max_msg_size. Note the MPICH eager/rendezvous threshold is independent of any internal libfabric provider threshold.
```

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h:50]: That sounds fine.

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_send.h]: Applied.

**Review by raffenet — APPROVED** [review]
I think this is OK to go. Just looking at it again, I'm wondering if we can convert the ack messages to use the `am_send_hdr` API rather than tsend/trecv, but thats minor.

**Comment by hzhou** [comment]
> I think this is OK to go. Just looking at it again, I'm wondering if we can convert the ack messages to use the `am_send_hdr` API rather than tsend/trecv, but thats minor.

We can't because the receiver does not have the sender request pointer. We could if we manage to send `sreq` in CQ meta data. But again that will add more complexity than the current tag method. Once sender received the `rreq` from CTS, additional protocol handshakes can be carried using regular AM methods.

**Pull request merged**